File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: (recipe | ingredient)[] = [];

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  // TODO: implement me
  let retString = recipeName.replace(/[-_]/g, ' ');
  retString = retString.replace(/[^a-zA-Z\s]/g, '');
  retString = retString.toLowerCase();
  if (retString.length === 0) {
    return null;
  }
  const wordsArray = retString.split(/\s+/);
  for (let i = 0; i < wordsArray.length; i++) {
    wordsArray[i] = wordsArray[i][0].toUpperCase() + wordsArray[i].slice(1);
  }
  retString = wordsArray.join(' ');

  return wordsArray.join(' ');
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req:Request, res:Response) => {
  // TODO: implement me
  const entry = req.body;
  if (entry.type !== 'recipe' && entry.type !== 'ingredient') {
    return res.status(400).json({ error: 'Invalid type' });
  } 

  if (cookbook.map(item => item.name).includes(entry.name)) {
    return res.status(400).json({ error: 'entry with that name already exists' });
  }

  if (entry.type == "recipe") {
    return addRecipe(entry, res);
  } else if (entry.type == "ingredient") {
    return addIngredient(entry, res);
  }
});

// Task 2 Helpers
const addRecipe = (entry: recipe, res: Response) => {
  const requiredItems = new Set();
  for (const item of entry.requiredItems) {
    if (requiredItems.has(item.name)) {
      return res.status(400).json({ error: 'more than one element per name in requiredItems' });
    } else {
      requiredItems.add(item.name);
    }
  }

  const newRecipe: recipe = { 
    type: entry.type, 
    name: entry.name,
    requiredItems: entry.requiredItems,
  };

  cookbook.push(newRecipe);
  return res.status(200).json({});
}

const addIngredient = (entry: ingredient, res: Response) => {
  if (entry.cookTime < 0) {
    return res.status(400).json({ error: 'Invalid cookTime' });
  }

  const newIngredient: ingredient = {
    type: entry.type,
    name: entry.name,
    cookTime: entry.cookTime,
  };

  cookbook.push(newIngredient);
  return res.status(200).json({});
}

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req:Request, res:Request) => {
  // TODO: implement me
  const targetRecipeName = req.query.name as string;

  const recipeIndex = cookbook.findIndex(recipe => recipe.name === targetRecipeName);
  if (recipeIndex === -1) {
    return res.status(400).json({ error: 'recipe with given name not found' });
  }

  if (cookbook[recipeIndex].type === 'ingredient') {
    return res.status(400).json({ error: 'searched name is not name of recipe' });
  }

  // for bfs, treating recipes as parent nodes and their requiredItems as child nodes -> leaf nodes are basically ingredients 
  let totalCookTime = 0;
  const ingredientMap = new Map();
  const queue = [];
  queue.push(cookbook[recipeIndex].name);
  
  while (queue.length !== 0) {
    let tempName = queue.shift();
    const tempIndex = cookbook.findIndex(entry => entry.name === tempName);

    if (tempIndex === -1) {
      return res.status(400).json({ error: 'recipe contains recipe/ingredient not in cookbook' });
    }

    if (cookbook[tempIndex].type === 'ingredient') {
      const currIngredient = cookbook[tempIndex] as ingredient;
      totalCookTime += currIngredient.cookTime;
      if (ingredientMap.has(cookbook[tempIndex].name)) {
        ingredientMap.set(currIngredient.name, 1 + ingredientMap.get(currIngredient.name));
      } else {
        ingredientMap.set(cookbook[tempIndex].name, 1);
      }
    }

    if (cookbook[tempIndex].type === 'recipe') {
      let currRecipe = cookbook[tempIndex] as recipe;
      
      for (const requiredItem of currRecipe.requiredItems) {
        for (let i = 0; i < requiredItem.quantity; i++) {
          queue.push(requiredItem.name);
        }
      }
    }

  }
  
  const ingredientsList = []
  for (const [itemName, numItem] of ingredientMap) {
    ingredientsList.push({ 
      name: itemName,
      quantity: numItem,
    });
  }

  const recipeSummary = {
    name: targetRecipeName,
    cookTime: totalCookTime,
    ingredients: ingredientsList,
  };

  return res.status(200).json(recipeSummary);
});

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
    name: str
    type: str

@dataclass
class RequiredItem():
    name: str
    quantity: int

@dataclass
class Recipe(CookbookEntry):
    required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
    cook_time: int


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook: List[Union[Recipe, Ingredient]] = []

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
    data = request.get_json()
    recipe_name = data.get('input', '')
    parsed_name = parse_handwriting(recipe_name)
    if parsed_name is None:
        return 'Invalid recipe name', 400
    return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> Union[str | None]:
    # TODO: implement me
    ret_string = re.sub(r'[-_]', ' ', recipeName.lower())
    ret_string = re.sub(r'[^a-zA-Z\s]', '', ret_string)
    
    if len(ret_string) == 0:
        return None

    words_array = ret_string.split()

    return " ".join(words_array).title()

# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
    # TODO: implement me
    entry = None
    dict_entry = request.get_json()
    
    for item in cookbook:
        if item.name == dict_entry['name']:
            return jsonify({ "error": 'entry with that name already exists' }), 400
    
    if dict_entry['type'] == 'recipe':
        required_items = []
        for required_item in dict_entry['requiredItems']:
            required_items.append(RequiredItem(
                name = required_item['name'],
                quantity = required_item['quantity']
            ))
        entry = Recipe(
            name = dict_entry['name'],
            type = dict_entry['type'],
            required_items = required_items,
        )
    elif dict_entry['type'] == 'ingredient':
        entry = Ingredient(
            name = dict_entry['name'],
            type = dict_entry['type'],
            cook_time = dict_entry['cookTime'],
        )
    else:
        return jsonify({ "error": 'Invalid type' }), 400

    

    if entry.type == 'recipe':
        return add_recipe(entry)
    elif entry.type == 'ingredient':
        return add_ingredient(entry)

# task 2 helpers
def add_recipe(entry):
    required_items_set = set()
    for item in entry.required_items:
        if item.name in required_items_set:
            return jsonify({ "error": 'more than one element per name in requiredItems' }), 400
        else:
            required_items_set.add(item.name)

    cookbook.append(entry)
    return jsonify({}), 200

def add_ingredient(entry):
    if entry.cook_time < 0:
        return jsonify({ "error": 'Invalid cooktime' }), 400

    cookbook.append(entry)
    return jsonify({}), 200


# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
    target_recipe_name = request.args.get('name')

    recipe_names = [entry.name for entry in cookbook]
    
    if target_recipe_name not in recipe_names:
        return jsonify({ "error": 'recipe with given name not found' }), 400
    recipe_index = recipe_names.index(target_recipe_name)

    if cookbook[recipe_index].type == 'ingredient':
        return jsonify({ "error": 'searched name is not name of recipe' }), 400
    
    total_cook_time = 0
    ingredient_dict = {}
    queue = []
    queue.append(cookbook[recipe_index].name)

    while len(queue) != 0:
        temp_name = queue.pop(0)
        if temp_name not in recipe_names:
            return jsonify({ "error": 'recipe contains recipe/ingredient not in cookbook' + temp_name }), 400

        temp_index = recipe_names.index(temp_name)

        if cookbook[temp_index].type == 'ingredient':
            curr_ingredient = cookbook[temp_index]
            total_cook_time += curr_ingredient.cook_time
            ingredient_dict[curr_ingredient.name] = 1 + ingredient_dict.get(curr_ingredient.name, 0)
        
        if cookbook[temp_index].type == 'recipe':
            curr_recipe = cookbook[temp_index]

            for required_item in curr_recipe.required_items:
                for i in range(required_item.quantity):
                    queue.append(required_item.name)

    ingredients_list = []
    for item_name, num_item in ingredient_dict.items():
        ingredients_list.append({
            "name": item_name,
            "quantity": num_item
        })
    
    recipeSummary = {
        "name": target_recipe_name,
        "cookTime": total_cook_time,
        "ingredients": ingredients_list
    }

    return jsonify(recipeSummary), 200


# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
    app.run(debug=True, port=8080)
